### ISSUE TITLE
Support supplying flags like Prune in the app updater when requesting a sync from ArgoCD

### ISSUE THREAD

**Issue by andrii-korotkov-verkada** [akuity/kargo#2984]
# Checklist

* [x] I've searched the issue queue to verify this is not a duplicate feature request.
* [ ] I've pasted the output of `kargo version`, if applicable.
* [ ] I've pasted logs, if applicable.

# Proposed Feature

ArgoCD can pass extra options to sync, e.g. Prune https://github.com/argoproj/argo-cd/blob/32cc6638f78e15652c24eb2cd6403e6e7ec26ab0/server/application/application.go#L1971. Kargo doesn't set Prune, yet sets SyncOptions https://github.com/akuity/kargo/blob/cc293f0bd128943e097d9ae76ea69469b7a1c462/internal/directives/argocd_updater.go#L511-L528. It be nice to allow specifying Prune with update requests.

# Motivation

Parity with ArgoCD sync.

# Suggested Implementation

Surface Prune checkbox in UI if not already, then pass the flag to the sync field in the operation.

**Comment by krancour** [comment]
> flags like Prune

Prune seems like a sensible option to expose. If there are others you feel would be especially useful, please enumerate.

I don't want to blindly expose _all_ options, as my gut says not all of them make sense in the context of Kargo.

Prune, at least, is absolutely doable.

**Comment by andrii-korotkov-verkada** [comment]
There also can be Replace, Force, Dry Run. Regarding Replace, there's this proposal https://github.com/argoproj/argo-cd/issues/20730, that would be discussed soon.

**Comment by krancour** [comment]
Rather than exposing any of these as options in the `argocd-update` step, I'm wondering if these are just things that should be copied over from the Application's sync policy when we create the sync operation.

Is there any reason that the options used by `argocd-update` should differ from what's in the sync policy?

**Comment by jessesuen** [comment]
> I'm wondering if these are just things that should be copied over from the Application's sync policy when we create the sync operation.

Actually, I already assumed we were doing that, so we should make that change. I believe Argo CD auto-sync will pick up the applications preferences when creating the operation, so we can copy that logic.

**Comment by krancour** [comment]
> Actually, I already assumed we were doing that, so we should make that change.

I thought there were, perhaps, things we ought to be copying but weren't but I just double-checked and we _are_ already copying all sync options.

So @andrii-korotkov-verkada, if you have specific sync options set on the App, Kargo will already use them.

> Sync flags might be good to support in the argocd-update step too, in case there is a use case to override the Application's default sync settings.

> I believe because sync options are just generic string, we just need to support a string field and pass it on. There would be no upkeep of supporting more and more options.

I agree, especially if it's this straightforward.

Thanks @jessesuen.

**Comment by am-ltk** [comment]
It doesn't look like there is a way to set "prune" flag on the step as it's not a SyncOption and there is no SyncOption equivalent for that: https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/types.go#L1380-L1386

It would be great to know if there is a workaround or get this implemented in the schema

**Comment by krancour** [comment]
This probably should have been closed. I will double check in the coming week, but to my best recollection that step is already copying options from the App it is syncing, which would obviate the need for surfacing such options on the step itself.

**Comment by am-ltk** [comment]
> probably should have been closed. I will double check in the coming week, but to my best recollection that step is already copying options from the App it is syncing, which would obviate the need for surfacing such options on the step itself.

@krancour Yes, it is copying the "SyncOptions" but SyncOptions have no support for Prune for an on-demand sync that Kargo uses. I.e., you can't enable it the metadata of the Argo App, there is no "Prune=true" option. You can only set it for "auto" sync but not for on-demand ones. So, currently, all Kargo syncs have "Prune" disabled if you have auto sync off on the Argo side. We have it off for a reason as we wanted deploys to always be initiated from the Kargo side and not when Argo wants to reconcile something by itself.

**Comment by krancour** [comment]
@am-ltk ok. Somehow, I'd not realized that prune couldn't be set as a normal sync option and was only something you could choose at the time of initiating a manual sync. Let me just confirm that, and if that is indeed the case, I think we could probably surface prune as an option.

**Comment by am-ltk** [comment]
Wondering if you had a chance to confirm it on your side.

**Comment by am-ltk** [comment]
@jessesuen @krancour do you guys have any updates on this issue by any chance? Leaving deleted resources creates toil and confusion for developers.

**Comment by kamilkrampa** [comment]
@jessesuen @krancour I think previous comments regarding syncOptions are valid. I created a draft #6837. Can you take a look into this please?

**Comment by jessesuen** [comment]
Copying [my response from the PR](https://github.com/akuity/kargo/pull/6837#issuecomment-5588911955) here:

The team has been feeling for some time that `argocd-update` is getting overloaded. There are other sync options (such as selective sync) that warrant [a separate `argocd-sync` step](https://github.com/akuity/kargo/issues/6028). My preference is not to add these types of features to argocd-update and put them in a new argocd-sync instead.
